_posts/2017-06-15-REST-api-NodeJS-koa.md

```diff
@@ -0,0 +1,230 @@
+ 
+NOTE: this setup works well, but the starting of the Docker container takes ~30 seconds, which is quite long, considering that the tests take maybe a few seconds. In a continuous integration build, this doesn’t matter as much, but when you are trying to fix tests, it does take a lot of time if you have to wait about a minute for each test run.
+
+# Task runner
+
+### Webpack
+
+When I started this project, I was looking up some best practices for node. I came across an [article](http://jlongster.com/Backend-Apps-with-Webpack--Part-I) that suggested you should use [Webpack](https://webpack.github.io/) for backend too. I already have some experience with Webpack from frontend development, so at first it seemed logical to use it for backend too. When I was trying to get the dependency injection to work with awilix, I realized that I could not pass any paths to libraries, because when my code was bundled, the paths would be invalid. Then I started to actually wonder why I was bundling my code. In frontend you bundle your code to make it as small as possible so you don’t waste the user’s bandwidth and make you website load faster, but in backend, that does not matter, since the code does not have to be sent anywhere. At this point I decided I didn’t need Webpack at all and I could just use npm scripts functionality to create tasks.
+
+### npm
+
+Npm is actually the only build tool you need. If what you want to do is more than a single line command, you can just write scripts in either TypeScript (you can execute them with ts-node), bash, JavaScript, … whatever you like. I wrote my scripts in TypeScript, because to me it makes more sense to use TypeScript for everything, but I could just as well have written them using bash. Npm also gives you pre and post task hooks. So if you write a task with the name "e2e" as I did, you can also add a task with the "pre" prefix or the "post" prefix that will automatically be executed before and after the task is executed. This way I could easily separate the starting of the Docker container, the executing of the tests and the stopping of the Docker container into different scripts. I could then just execute npm run pree2e to check if my script to start the Docker worked. I really like this approach and the fact that I don’t need another tool to learn like gulp or Webpack.
```

Comment:
no, the convention is 'pre' + taskname, since my task name is e2e, you need to run pree2e. I don't think npm script names can have spaces in them.